app/src/main/java/org/p2p/wallet/user/repository/prices/impl/TokenPricesCoinGeckoRepository.kt

```diff
@@ -36,19 +39,23 @@ class TokenPricesCoinGeckoRepository(
 
     private suspend fun loadPrices(tokenIds: List<TokenId>, targetCurrencySymbol: String): List<TokenPrice> =
         withContext(dispatchers.io) {
-            val tokenIdsForRequest = tokenIds.joinToString(",") { it.id }
-            try {
-                coinGeckoApi.getTokenPrices(
-                    tokenIds = tokenIdsForRequest,
-                    targetCurrency = targetCurrencySymbol.lowercase()
-                )
-                    .map { TokenPrice(tokenId = it.id, price = it.currentPrice) }
-            } catch (httpException: HttpException) {
-                val errorCode = httpException.code()
-                if (errorCode == 429 || errorCode == 403) {
-                    Timber.e(RequestRateLimitMet(httpException))
+            val filteredTokenIds = priceCacheRepository.filterKeysForExpiredPrices(tokenIds.map { it.id })
+            if (filteredTokenIds.isNotEmpty()) {
+                val tokenIdsForRequest = tokenIds.joinToString(",") { it.id }
+                try {
+                    val response = coinGeckoApi.getTokenPrices(
+                        tokenIds = tokenIdsForRequest,
+                        targetCurrency = targetCurrencySymbol.lowercase()
+                    ).associate { it.id to TokenPriceWithMark(it.currentPrice) }
+                    priceCacheRepository.mergeCache(response)
+                } catch (httpException: HttpException) {
+                    val errorCode = httpException.code()
+                    if (errorCode == 429 || errorCode == 403) {
+                        Timber.e(RequestRateLimitMet(httpException))
+                    }
+                    throw httpException
                 }
-                throw httpException
             }
+            priceCacheRepository.getPrices().entries.map { TokenPrice(tokenId = it.key, price = it.value.priceInUsd) }
```

Comment:
can you add ubit tests please?

Reply:
> can you add unit tests please?
i do not even know what tests i can write here?
check that cache working? by timestamp?

Reply:
you can also add check that we fetch correct popular tokens and it's prices

Reply:
we need to assign Test Guru on this test =) @gslevinkov 😈